feat: add minimal `atlas` step to the build FC-0012

OmarIthawi opened the pull request:
Add https://github.com/openedx/openedx-atlas to the Dockerfile and run the `atlas pull` command on every build.

I'm starting here because this is a simple plugin and helps us to design it properly. The end goal is to add `atlas` on all other components including edX Platform, MFEs and possibly even ecommerce.

This is a very minimal build that needs the following to be viable:

 - Ability to configure the repo, branch and other options.
 - Possibly make it a system-wide pattern in tutor instead of adding it to individual components.

This PR requires the following branch of discovery: https://github.com/openedx/course-discovery/pull/4037


### TODO

 - [ ] ~Implement https://github.com/openedx/openedx-atlas/issues/17~
 - [ ] ~Implement https://github.com/openedx/openedx-atlas/issues/30~
 - [x] Update the version in this repo
 - [x] Add a single `DISCOVERY_ATLAS_OPTOINS` variable to keep it minimal
 - [x] Build the docker image and test 
 - [x] Ensure messages gets compiled correctly: https://github.com/openedx/openedx-translations/pull/564

<details><summary>Building and testing logs</summary>

```
# Set the following variables:
# DISCOVERY_ATLAS_ARGS: --filter=ar,de
# DISCOVERY_ATLAS_PULL: true

$ tutor config save
$ tutor images build discovery
Building image docker.io/overhangio/openedx-discovery:16.0.0
....
$ docker run -it docker.io/overhangio/openedx-discovery:16.0.0 atlas --version
v0.4.4
```

Check the file tree

```
$ docker run -it docker.io/overhangio/openedx-discovery:16.0.0 ls -R course_discovery/conf/locale/
course_discovery/conf/locale/:
ar  config.yaml  de

course_discovery/conf/locale/ar:
LC_MESSAGES

course_discovery/conf/locale/ar/LC_MESSAGES:
djangojs.po  django.po

course_discovery/conf/locale/de:
LC_MESSAGES

course_discovery/conf/locale/de/LC_MESSAGES:
djangojs.po  django.po
```

Verify translations within the image:

```
$ docker run -it docker.io/overhangio/openedx-discovery:16.0.0 cat course_discovery/conf/locale/de/LC_MESSAGES/django.po | head -n 100
# #-#-#-#-#  django.po (course-discovery)  #-#-#-#-#
# edX translation file
# Copyright (C) 2018 edX
# This file is distributed under the GNU AFFERO GENERAL PUBLIC LICENSE.
# 
# Translators:
# Translators:
# Omar Al-Ithawi <i@omardo.com>, 2023
# 
msgid ""
msgstr ""
"Project-Id-Version: edx-platform\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2023-08-03 00:22+0000\n"
"PO-Revision-Date: 2023-01-26 15:39+0000\n"
"Last-Translator: Omar Al-Ithawi <i@omardo.com>, 2023\n"
"Language-Team: German (https://app.transifex.com/open-edx/teams/147691/de/)\n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Language: de\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"

#: apps/api/filters.py:47
#, python-brace-format
msgid "No user with the username [{username}] exists."
msgstr "Ein Nutzer mit Namen [{username}] existiert nicht."

#: apps/api/filters.py:50
msgid ""
"Only staff users are permitted to filter by username. Remove the username "
"parameter."
msgstr ""
"Nur Mitarbeiter sind berechtigt nach Benutzernamen zu filtern. Entfernen Sie"
" den Benutzername Parameter"

#: apps/api/serializers.py:813
msgid "Number of courses contained in this catalog"
msgstr "Anzahl der Kurse in diesem Katalog"

#: apps/api/serializers.py:816
msgid "Usernames of users with explicit access to view this catalog"
msgstr ""
"Benutzernamen der Nutzer mit ausdrücklichem Zugriff um diesen Katalog zu "
"sehen"

#: apps/api/serializers.py:976
msgid "Start date cannot be after the End date"
msgstr "Startdatum kann nicht nach dem Enddatum gesetzt werden"

#: apps/api/serializers.py:981
msgid "Term cannot be changed"
msgstr "Begriff kann nicht geändert werden"

#: apps/api/serializers.py:993
msgid "Language in which the course is administered"
msgstr "Sprache in der dieser Kurs verwaltet wird"

.....

```
</details> 

# Background
This contribution is part of the [FC-0012 project](https://openedx.atlassian.net/l/cp/XGS0iCcQ) which is sparked by the [Translation Infrastructure update OEP-58](https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0058-arch-translations-management.html#specification).

regisb commented on tutordiscovery/plugin.py:
```diff
@@ -32,6 +32,8 @@
         "CACHE_REDIS_DB": "{{ OPENEDX_CACHE_REDIS_DB }}",
         "EXTRA_PIP_REQUIREMENTS": [],
+        "ATLAS_VERSION": "0.1.0",
+        "ATLAS_PULL": True,
```
Let's not introduce these new settings, they are not necessary.

regisb commented on tutordiscovery/templates/discovery/build/discovery/Dockerfile:
```diff
@@ -42,6 +42,11 @@ RUN pip install nodeenv==1.7.0
 ENV PATH /openedx/nodeenv/bin:${PATH}
 
+# Install the Open edX Atlas translation tool
+RUN curl https://github.com/openedx/openedx-atlas/releases/{{ DISCOVERY_ATLAS_VERSION }}/atlas \
```
- We can hardcode the version to 0.1.0.
- Do we need to add the `-L` option?

OmarIthawi replied:
yes

regisb commented on tutordiscovery/templates/discovery/build/discovery/Dockerfile:
```diff
@@ -62,7 +67,10 @@ RUN {% if is_buildkit_enabled() %}--mount=type=cache,target=/openedx/.cache/pip,
-# Collect static assets
+# Update i18n and collect static assets
+{% if DISCOVERY_ATLAS_PULL %}
+RUN make OPENEDX_ATLAS_PULL=yes ATLAS_PULL_LANGS="{{ DISCOVERY_ATLAS_PULL_LANGS }}" pull_translations
```
- The `DISCOVERY_ATLAS_PULL_LANGS` setting is undefined. How can we make Atlas automatically (or at least: by default) download strings from all "sufficiently translated" languages? As a rule of thumb, we like to include in Tutor all languages with >50% reviewed strings.
- In Palm, the `make pull_translations` command is going to run `tx`, and not `atlas`. Actually, neither does the master branch run Atlas. 
- Also, how are we going to pull translations from Transifex (I assume?) if there is no transifexrc file in the Docker image?
- Actually, do we really have to run a `make` command? Can we just run `atlas` directly to remove one level of indirection?
- Once we have a basic version working, we should try to move the atlas step to a separate Docker layer.

regisb reviewed (changes requested):
I like your approach of testing Palm on a smaller Tutor plugin. And I am also looking forward to ditching my very custom solution based on [openedx-i18n](https://github.com/openedx/openedx-i18n/). Let's try to get Atlas working on this repo before we attempt to tackle the openedx Dockerfile.

I'm curious, did you manage to build the discovery image with this PR? If yes, did you run your own fork of course-discovery? I did not test your changes but I suspect that I would not be able to build the Docker image without transifex credentials.

regisb commented:
I converted this PR to a draft, let me know when it's ready for another round of review.

OmarIthawi commented:
@regisb I've addressed some of your notes.

I think we still need some way to pass options for example: 
 - Pull from another repo.
 - Specify which languages to pull.
 - Other options that atlas allows: https://github.com/openedx/openedx-atlas/pull/28/files#diff-7b3ed02bc73dc06b7db906cf97aa91dec2b2eb21f2d92bc5caa761df5bbc168fR59

Another feature is that `atlas pull` should be disabled by default, because at the moment it pulls an incomplete translation and it's worse than what's checkout on GitHub.


### Atlas and Open edX Translations
Perhaps, it's good to share a recap on what happened on Atlas. Part of OEP-58, we've created two new additions to the Open edX Platform https://github.com/openedx/openedx-translations and https://github.com/openedx/openedx-atlas .

`atlas pull` replaces `tx pull` in which that it doesn't pull from Transifex, but pulls from the https://github.com/openedx/openedx-translations GitHub repo which makes it possible to pull without needing authentication or `.transifexrc` file.

### Testing

```
$ tutor images build discovery
$ tutor dev run discovery -- atlas --help  # Prints the help message
```

### File tree
The file tree seems about right. It removed the old files and add fresh ones from the repository (as opposed to Transifex).

```
$ tutor dev run discovery -- ls -R course_discovery/conf/locale
course_discovery/conf/locale:
ar  config.yaml  de  en  fr_CA

tutor dev run discovery -- ls -R course_discovery/conf/locale
course_discovery/conf/locale/ar:
LC_MESSAGES

course_discovery/conf/locale/ar/LC_MESSAGES:
djangojs.po  django.po

course_discovery/conf/locale/de:
LC_MESSAGES

course_discovery/conf/locale/de/LC_MESSAGES:
djangojs.po  django.po

course_discovery/conf/locale/en:
LC_MESSAGES

course_discovery/conf/locale/en/LC_MESSAGES:
djangojs.po  django.po

course_discovery/conf/locale/fr_CA:
LC_MESSAGES

course_discovery/conf/locale/fr_CA/LC_MESSAGES:
djangojs.po  django.po
```

brian-smith-tcril commented:
I think we should wait until we implement https://github.com/openedx/openedx-atlas/issues/34 and use pip

OmarIthawi commented on tutordiscovery/templates/discovery/build/discovery/Dockerfile:
```diff
@@ -60,9 +60,18 @@ RUN {% if is_buildkit_enabled() %}--mount=type=cache,target=/openedx/.cache/pip,
+    openedx-atlas'<1.0.0'
+
+# Update i18n and collect static assets
+{% if DISCOVERY_ATLAS_PULL %}
```
Pulls all languages. 

Your point that atlas should pull only relatively complete languages is probably a feature in `atlas` which I've added as a suggestion:

 - https://github.com/openedx/openedx-atlas/issues/44

OmarIthawi commented:
@regisb and @brian-smith-tcril would you mind taking a look at this PR? It should be ready again for review.

regisb commented on tutordiscovery/plugin.py:
```diff
@@ -32,6 +32,8 @@
         "CACHE_REDIS_DB": "{{ OPENEDX_CACHE_REDIS_DB }}",
         "EXTRA_PIP_REQUIREMENTS": [],
+        "ATLAS_PULL": False,
+        "ATLAS_ARGS": "",
```
Don't add these configuration settings. Assume all users should run Atlas by default. Why wouldn't they?

OmarIthawi replied:
> Don't add these configuration settings. Assume all users should run Atlas by default. Why wouldn't they?

I understand the argument that `ATLAS_PULL` can be removed and enabled by default.

However, suppose a engineer wants to select which languages atlas will pull how should we go about that?
For example, we need `atlas pull` to have the `--filter=ar,fr_CA` argument on my build. Without  `ATLAS_ARGS`, I'd need to fork the `tutor-discovery` plugin and edit the Dockerfile or maybe add a copy of the Dockerfile template in my local configs.

I'm assuming we could use a global tutor-wide `ATLAS_ARGS` if this helps to reduce the need for forking.
We could alternatively use a `patch` placeholder for extensibility:

`atlas pull {{ patch("atlas-extra-args") }} ...`

The full atlas docs are available here: https://github.com/openedx/openedx-atlas#usage

What do you think?

OmarIthawi commented on tutordiscovery/templates/discovery/build/discovery/Dockerfile:
```diff
@@ -60,9 +60,18 @@ RUN {% if is_buildkit_enabled() %}--mount=type=cache,target=/openedx/.cache/pip,
-    uwsgi==2.0.21
+    uwsgi==2.0.21 \
+    # Open edX Atlas tranlsation management tool https://pypi.org/project/openedx-atlas/
+    openedx-atlas'<1.0.0'
```
I think we'll pin the version within `discovery` itself because it has the `pip-tools` setup for automatic upgrades.

OmarIthawi commented on tutordiscovery/templates/discovery/build/discovery/Dockerfile:
```diff
@@ -60,9 +60,18 @@ RUN {% if is_buildkit_enabled() %}--mount=type=cache,target=/openedx/.cache/pip,
+
+# Update i18n and collect static assets
+{% if DISCOVERY_ATLAS_PULL %}
+RUN find course_discovery/conf/locale -mindepth 1 -maxdepth 1 -type d -exec rm -r {} \; \
```
> My bash skills are poor... 

I doubt it 😅, nevertheless the command isn't very readable and I'll split it :)

OmarIthawi commented:
@regisb I've updated the PR except for the `args` part which I tend to think it's valuable but I'd like to hear from you.

Anyway, it's not a blocker for this PR so we can remove it until we figure out another alternative.


```
axim $ tutor config save
Configuration saved to /home/omar/work/axim/config.yml
Environment generated in /home/omar/work/axim/env


axim $ tutor images build discovery
Building image docker.io/overhangio/openedx-discovery:16.0.0
....
 => => naming to docker.io/overhangio/openedx-discovery:16.0.0


axim $ tutor dev run discovery bash

app@c91d22d98665:~/discovery$ ls -R course_discovery/conf/locale/
course_discovery/conf/locale/:
ar  config.yaml  de  en  fr_CA

course_discovery/conf/locale/ar:
LC_MESSAGES

course_discovery/conf/locale/ar/LC_MESSAGES:
djangojs.po  django.po

course_discovery/conf/locale/de:
LC_MESSAGES

course_discovery/conf/locale/de/LC_MESSAGES:
djangojs.po  django.po

course_discovery/conf/locale/en:
LC_MESSAGES

course_discovery/conf/locale/en/LC_MESSAGES:
djangojs.po  django.po

course_discovery/conf/locale/fr_CA:
LC_MESSAGES

course_discovery/conf/locale/fr_CA/LC_MESSAGES:
djangojs.po  django.po
```

OmarIthawi commented:
@regisb what do you think should happen if a `po` file fails to compile due to Translator writing invalid strings `Hi {user}` --> `Merhaba {kullanci}`?

Currently the docker build fails, which I think isn't appropriate.

I'm thinking we should add validation to the https://github.com/openedx/openedx-translations so this won't happen inside Tutor, correct?

OmarIthawi commented:
@regisb what do you think should happen if a `po` file fails to compile due to Translator writing invalid strings `Hi {user}` --> `Merhaba {kullanci}`?

Currently the docker build fails, which I think isn't appropriate.

I'm thinking we should add validation to the https://github.com/openedx/openedx-translations so this won't happen inside Tutor, correct?


cc: @brian-smith-tcril

brian-smith-tcril commented:
@OmarIthawi I made an issue on the `openedx-translations` repo for validation https://github.com/openedx/openedx-translations/issues/549

OmarIthawi commented:
Thanks Brian and Regis. I'll fix the openedx-translations repo first then I'll continue with this PR.

regisb commented on tutordiscovery/templates/discovery/build/discovery/Dockerfile:
```diff
@@ -60,7 +60,14 @@ RUN {% if is_buildkit_enabled() %}--mount=type=cache,target=/openedx/.cache/pip,
+    openedx-atlas==0.4.4
+
+# Update i18n: the equivalent of "make pull_translations"
+RUN find course_discovery/conf/locale -type d -mindepth 1 -maxdepth 1 -exec rm -rf {} + # Remove stale translation files
```
This causes all translation files currently in course_discovery/conf/locale to be deleted. Not all translation files will be replaced by `atlas pull`. This means that some languages will no longer be available to existing users, which is not acceptable.

Is there an alternate solution? Can we just skip the deletion step?

regisb replied:
> This ideally shouldn't be a concern as long as both projects translations have the same completeness. Which is still work in progress.

OK. So this PR is not 100% ready to be merged, right?

> What if some translations are broken in the original repo? Then atlas pull won't fix those errors. What do you think?

I'm fine with existing bugs, I just don't want to create new ones. It's not atlas' job to fix existing translations issues.

> In MFEs, it's preferable to have only the used languages to reduce the JavaScript bundle because translations are compiled and bundled into the code. But this isn't an MFE.

That's a good point. But then again, it's not like MFE bundles are suddenly going to get very big. They already contain all the translation strings, right? Plus, it seems to me that translations strings take little size compared to vendor libraries.

OmarIthawi replied:
> That's a good point. But then again, it's not like MFE bundles are suddenly going to get very big. They already contain all the translation strings, right? Plus, it seems to me that translations strings take little size compared to vendor libraries.

This is still an open discussion in the Transifex Working Group. But yeah, it shouldn't be a concern in this context.

OmarIthawi replied:
@regisb. It's ready'ish, so no it's not ready. I just saw your comment 😿 

I'll check with the team on the plugin decision and get back to you. Thanks.

regisb replied:
I'm eager to see Atlas succeed and don't want to cause extra delays. If it's a chicken and egg thing, and building a plugin would further delay the project, then we can help resolve the situation by hiding the feature behind an undocumented feature flag in the tutor-discovery plugin, just like you suggested. We should make it clear that it's a temporary, unsupported feature flag, and the flag should not be present in other core plugins.

Thus I no longer have any objection to your PR and I'd say I'm ready to merge it when you are.

OmarIthawi replied:
To be clear, making it a plugin isn't a bad choice, in fact it's pretty good one. However, it's bit more costly than direct pull requests, so I'd like to hear from @brian-smith-tcril.

I imagine if we have an `tutor-atlas` plugin which adds Tutor patches to all other plugins would be a nice way to make atlas optional.

> the feature behind an undocumented feature flag in the tutor-discovery plugin, just like you suggested

Great, thanks for being open for the lesser option for the sake of timeline. Today, we'll know which way to go.

regisb reviewed (changes requested):
In addition to my comment, can you please add a changelog entry with `scriv create`?

OmarIthawi commented:
@regisb the pull request is ready again for review.

OmarIthawi commented:
@regisb a gentile reminder about this pull request :)

regisb commented:
pulling in @Faraz32123 who is the new maintainer for this plugin.

OmarIthawi commented:
@Faraz32123 would you mind checking this pull request?

OmarIthawi commented:
@Faraz32123 would you mind checking this pull request?

regisb commented:
@OmarIthawi did you see my [last comment above](https://github.com/overhangio/tutor-discovery/pull/41#discussion_r1330057933)?

OmarIthawi commented:
@regisb Let's please continue with the feature flag.

I don't fully understand Tutor branching strategy, so should I change the base branch from `master` --> `nightly`?

regisb commented:
> I don't fully understand Tutor branching strategy, so should I change the base branch from master --> nightly?

Do you want the feature in the Palm release? If yes, then keep the base branch as master, and it will also be included in nightly. If the feature should only target the upstream master branch of the course-discovery repo, aim for the nightly branch here. For more information: https://docs.tutor.overhang.io/tutorials/nightly.html

OmarIthawi commented:
@regisb for the discovery, I think it should go to `nightly` so it's included in Quince.

All other `atlas` pull requests should go to `nightly` except for the MFE because the `communications` MFE needs `atlas` in Palm.

regisb commented:
> @regisb for the discovery, I think it should go to nightly so it's included in Quince.

To be clear: if atlas goes to the plugin master branch, the change will also be included in Quince. We should push to nightly only if the change does not work in Palm.

Does that make sense?

OmarIthawi commented:
> > @regisb for the discovery, I think it should go to nightly so it's included in Quince.
> 
> To be clear: if atlas goes to the plugin master branch, the change will also be included in Quince. We should push to nightly only if the change does not work in Palm.
> 
> Does that make sense?

Yes, makes sense :)

OmarIthawi commented:
@regisb @Faraz32123 I've rebased over `nightly` and tested the build and it pulled the translations correctly:

```
$ pip install -e tutor-discovery
$ tutor config save && tutor images build discovery

...
 => [minimal 16/19] RUN atlas pull  translations/course-discovery/course_discovery/conf/locale:course_discovery/conf/locale                           2.0s 
...

app@e4c97e0af60c:~/discovery/course_discovery/conf/locale$ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   ar/LC_MESSAGES/django.mo
	modified:   ar/LC_MESSAGES/django.po
	modified:   ar/LC_MESSAGES/djangojs.mo
	modified:   ar/LC_MESSAGES/djangojs.po
	modified:   da/LC_MESSAGES/djangojs.mo
	modified:   da/LC_MESSAGES/djangojs.po
	modified:   de/LC_MESSAGES/django.mo
	modified:   de/LC_MESSAGES/django.po
	modified:   de/LC_MESSAGES/djangojs.mo
	modified:   de/LC_MESSAGES/djangojs.po
	modified:   de_DE/LC_MESSAGES/djangojs.mo
	modified:   de_DE/LC_MESSAGES/djangojs.po
	modified:   el/LC_MESSAGES/djangojs.mo
	modified:   el/LC_MESSAGES/djangojs.po
	modified:   en/LC_MESSAGES/django.po
	modified:   en/LC_MESSAGES/djangojs.po
	modified:   fr_CA/LC_MESSAGES/django.mo
	modified:   fr_CA/LC_MESSAGES/django.po
	modified:   fr_CA/LC_MESSAGES/djangojs.mo
	modified:   fr_CA/LC_MESSAGES/djangojs.po
```

regisb reviewed (approved):
Thanks for bearing with me Omar!

OmarIthawi commented:
Hey!! That's a great way to end the day. Thanks @regisb!

> Thanks for bearing with me Omar!

On the contrary! It's been a great review. I'm new to Tutor anyway, so I'm expecting a lot of comments anyway.

shadinaif commented on tutordiscovery/plugin.py:
```diff
@@ -31,6 +31,7 @@
         "OAUTH2_KEY_SSO": "discovery-sso",
         "OAUTH2_KEY_SSO_DEV": "discovery-sso-dev",
         "CACHE_REDIS_DB": "{{ OPENEDX_CACHE_REDIS_DB }}",
+        "ATLAS_PULL": False,
```
@OmarIthawi , shouldn't this be `OPENEDX_ATLAS_PULL` ?

OmarIthawi replied:
@shadinaif the prefix is automatically added via Tutor. So in the `config.yaml` file it will be `DISCOVERY_ATLAS_PULL: true`.

shadinaif replied:
Thank you!
